lib/Dialect/FIRRTL/Ops.cpp

```diff
@@ -699,6 +700,93 @@ FIRRTLType MemOp::getDataTypeOrNull() {
+  bool srcIsFlip = srcElementType.isa<FlipType>();
+
+  if (destElementName != srcElementName)
+    return false;
+
+  if (destIsFlip != srcIsFlip)
+    return false;
+
+  // If either destination or source is flipped, strip the outer flip for
+  // recursively equivalance checking.
+  if (destIsFlip)
+    destElementType = destElementType.cast<FlipType>().getElementType();
+
+  if (srcIsFlip)
+    srcElementType = srcElementType.cast<FlipType>().getElementType();
+
+  return areConnectEquivalent(destElementType, srcElementType);
+}
+
+/// Returns whether the two types are equivalent for the purpose of
+/// connections. See Section 4.5 in the FIRRTL spec:
+/// https://github.com/freechipsproject/firrtl/blob/master/spec/spec.pdf.
+static bool areConnectEquivalent(FIRRTLType destType, FIRRTLType srcType) {
+  // Analog types cannot be connected and must be attached.
+  if (destType.isa<AnalogType>() || srcType.isa<AnalogType>())
+    return false;
+
+  // Reset types can be driven by UInt<1>, AsyncReset, or Reset types.
+  if (destType.getPassiveType().isa<ResetType>())
+    return srcType.isResetType();
+
+  // Reset types can drive UInt<1>, AsyncReset, or Reset types.
+  if (srcType.isa<ResetType>())
+    return destType.getPassiveType().isResetType();
+
+  // Vector types can be connected if they have the same size and element type.
+  auto destVectorType = destType.dyn_cast<FVectorType>();
+  auto srcVectorType = srcType.dyn_cast<FVectorType>();
+  if (destVectorType && srcVectorType)
+    return destVectorType.getNumElements() == srcVectorType.getNumElements() &&
+           areConnectEquivalent(destVectorType.getElementType(),
+                                srcVectorType.getElementType());
+
+  // Bundle types can be connected if they have the same size, element names,
+  // and element types.
+  auto destBundleType = destType.dyn_cast<BundleType>();
+  auto srcBundleType = srcType.dyn_cast<BundleType>();
+  if (destBundleType && srcBundleType)
+    return destBundleType.getNumElements() == srcBundleType.getNumElements() &&
+           llvm::all_of(llvm::zip(destBundleType.getElements(),
+                                  srcBundleType.getElements()),
+                        areElementsConnectEquivalent);
+
+  // Ground types can be connected if their passive, widthless versions are
+  // equal.
+  return destType.getPassiveType().getWidthlessType() ==
+         srcType.getPassiveType().getWidthlessType();
+}
+
+static LogicalResult verifyConnectOp(ConnectOp connect) {
+  FIRRTLType destType = connect.dest().getType().cast<FIRRTLType>();
+  FIRRTLType srcType = connect.src().getType().cast<FIRRTLType>();
+
+  if (!areConnectEquivalent(destType, srcType))
+    return connect.emitError("type mismatch between destination ")
+           << destType << " and source " << srcType;
+
+  // TODO(mikeurbach): verify source bitwidth is >= destination bitwidth.
+  // TODO(mikeurbach): verify destination flow is sink or duplex.
```

Comment:
I think that FIRRTL supports connecting to inputs.  Am I wrong about this?  I am pretty sure this occurred in real code, even if not documented.  @seldridge do you know?

Reply:
Connecting to source flow objects is disallowed based on "flow" checking, but not based on type checking. For ground types this means you cannot assign an input.

E.g., the following circuit will fail during the `CheckFlows` pass.

```
# cat Foo.fir
circuit Foo:
  module Foo:
    input in: UInt<1>
    in <= UInt<1>(0)
# firrtl -i Foo.fir
Exception in thread "main" firrtl.passes.CheckFlows$WrongFlow: : [module Foo]  Expression in is used as a SinkFlow but can only be used as a SourceFlow.
```

Note that this isn't a blanket "you can't connect to inputs" just that source flow can't be a LHS in an expression. If you make the input a bundle and put flips you can assign to the correctly flowing portions. E.g., the following is fine:

```
# cat Foo.fir
circuit Foo:
  module Foo:
    input in: { flip a: UInt<1> }
    in.a <= UInt<1>(0)
```